src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java

```diff
@@ -155,6 +186,71 @@ private Mono<Void> validateOpaqueReferenceToken(String issBaseUri, String token,
                 });
     }
 
+    /**
+     * Validates a token's audience or client ID depending on token type.
+     *
+     * JWT ID tokens (representing end users) typically contain an 'aud' claim that needs validation.
+     * JWT access tokens may not have an 'aud' claim but instead use a 'client_id' claim.
+     *
+     * We first try to validate the audience. Only if no valid audience is found, we fall back to
+     * client ID validation. This allows both token types to work with the gateway.
+     *
+     * IMPORTANT NOTES:
+     * - Currently, we only allow GridSuite audience tokens in the allowedAudiences configuration
+     * - If we want to allow other frontend applications to access this API:
+     *   a) This validation logic needs to be reviewed and possibly expanded
+     *   b) The CORS strategy would need to be modified accordingly to allow those origins
+     *
+     * @param jwtClaimsSet The JWT claims set
+     * @param exchange The server web exchange
+     * @return Mono<Void> with unauthorized response if validation fails, null if validation passes
+     */
+    private Mono<Void> validateAudienceOrClientId(JWTClaimsSet jwtClaimsSet, ServerWebExchange exchange) {
```

Comment:
Maybe better to return a synchronous value (boolean true false, or void/exception like the lib validator) and completeWithCode in the caller